<a href="https://colab.research.google.com/github/constantinembufung/deep_learning_with_keras/blob/master/Deep_Learning_with_tensorflow_recognizing_handwritten_digits.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MNIST with tensflow 2.0 and keras**

Today we will learn how to build a network that can recognize handwritten numbers. the dataset containts 60.000 examples for training and 10,000 samples for testing the model.

Train samples are annotated by humans with correct answers. 
We are going to do a [supervised learning ](https://en.wikipedia.org/wiki/Supervised_learning#:~:text=Supervised%20learning%20is%20the%20machine,a%20set%20of%20training%20examples.)since we have the correct answers.
Fig 1: show sample digit from the train dataset.
![alt text](https://static.packt-cdn.com/products/9781838823412/graphics/Images/B14761_01_14.png)

Each digit image is a grayscale image and consist of 28x28 pixels.

Since we have digits from 0 to 9, these are categorical data so we are going to use on-hot encoding to encode the digits to be able to feed them to our neural network. 
For example uisng OHE, 2 will be represented a [0,0,1,0,0,0,0,0,0,0] and 5 will be represented as [0,0,0,0,0,1,0,0,0,0]. OHE is very common tool used in data mining.

We will split our dataset into X_train, X_test. The data is converted to float32 to use 32-bit precision and normalized to the range[0,1] using activations functions.
More so, the true labels will be represented using Y_train, Y_test

In [3]:
import tensorflow as tf
import numpy as np
from tensorflow import keras
# Network and training parameters.
EPOCHS = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10   # number of outputs = number of digits
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION

In the above we have define EPOCH = 200, meaning we have to run the training in 200 steps
Batch_size is the number of samples you feed in to your network at a time and validation_split is data reserved for validating our training process


In [ ]:
mnist = keras.datasets.mnist
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()
# X_train is 60000, X_test is 10000 and all oshape 28x28. 
RESHAPED = 784 # we flattered the images
X_train.shape


(60000, 28, 28)

In [ ]:
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train.shape

(60000, 784)

In [ ]:
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# Normalize inputs to be within in [0, 1].
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')


60000 train samples
10000 test samples


In [ ]:
#one hot encoding of the labels
Y_train = tf.keras.utils.to_categorical(Y_train, NB_CLASSES)
Y_test = tf.keras.utils.to_categorical(Y_test, NB_CLASSES)

We are going to use the softmax activvation function. softmax functions outputs is in the range [0,1] when the inputs varies from -ve infinity to positive infinity. that is softmax squashes a k-dimensional vector int he range (0,1)

In [ ]:
# Build the model.
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(NB_CLASSES,
   input_shape=(RESHAPED,),
   name='dense_layer', 
   activation='softmax'))




we have define our model but now we need to compile the model by selecting a optimizer, a loss function and a metrice so  it can be executed in tensorflow: common objective functions includes: MSE, Binary_crossentropy, categrical_crossentropy

In [ ]:
# Compiling the model.
model.compile(optimizer='SGD', 
              loss='categorical_crossentropy',
              metrics=['accuracy'])


In [ ]:
# Training the model.
model.fit(X_train, Y_train,
               batch_size=BATCH_SIZE, epochs=EPOCHS,
               verbose=VERBOSE, validation_split=VALIDATION_SPLIT)


Epoch 1/200
375/375 [==============================] - 1s 2ms/step - loss: 0.2761 - accuracy: 0.9231 - val_loss: 0.2753 - val_accuracy: 0.9237
Epoch 2/200
375/375 [==============================] - 1s 2ms/step - loss: 0.2759 - accuracy: 0.9230 - val_loss: 0.2753 - val_accuracy: 0.9240
Epoch 3/200
375/375 [==============================] - 1s 2ms/step - loss: 0.2758 - accuracy: 0.9230 - val_loss: 0.2752 - val_accuracy: 0.9241
Epoch 4/200
375/375 [==============================] - 1s 2ms/step - loss: 0.2757 - accuracy: 0.9230 - val_loss: 0.2751 - val_accuracy: 0.9242
Epoch 5/200
375/375 [==============================] - 1s 2ms/step - loss: 0.2756 - accuracy: 0.9228 - val_loss: 0.2750 - val_accuracy: 0.9239
Epoch 6/200
375/375 [==============================] - 1s 2ms/step - loss: 0.2755 - accuracy: 0.9232 - val_loss: 0.2750 - val_accuracy: 0.9239
Epoch 7/200
375/375 [==============================] - 1s 2ms/step - loss: 0.2754 - accuracy: 0.9231 - val_loss: 0.2749 - val_accuracy: 0.9237

In [ ]:
#evaluate the model
test_loss, test_acc = model.evaluate(X_test, Y_test)
print('Test accuracy:', test_acc)


313/313 [==============================] - 0s 1ms/step - loss: 0.2709 - accuracy: 0.9242
Test accuracy: 0.9241999983787537


So, congratulations! You have just defined your first neural network in TensorFlow 2.0. 

#Improving the simple net in TensorFlow 2.0 with hidden layers
We had a test accuracy of 92.4% thats good. we now add hidden layers to our network to learn more complex patterns on training data. we also use ReLU activiation function. So we added 2 hidden layers

In [ ]:
import tensorflow as tf
from tensorflow import keras
# Network and training.
EPOCHS = 50
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10   # number of outputs = number of digits
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION
# Loading MNIST dataset.
# Labels have one-hot representation.
mnist = keras.datasets.mnist
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()
# X_train is 60000 rows of 28x28 values; we reshape it to 60000 x 784.
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# Normalize inputs to be within in [0, 1].
X_train, X_test = X_train / 255.0, X_test / 255.0
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# Labels have one-hot representation.
Y_train = tf.keras.utils.to_categorical(Y_train, NB_CLASSES)
Y_test = tf.keras.utils.to_categorical(Y_test, NB_CLASSES)
# Build the model.
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(N_HIDDEN,
          input_shape=(RESHAPED,),
          name='dense_layer', activation='relu'))
model.add(keras.layers.Dense(N_HIDDEN,
          name='dense_layer_2', activation='relu'))
model.add(keras.layers.Dense(NB_CLASSES,
          name='dense_layer_3', activation='softmax'))
# Summary of the model.
model.summary()
# Compiling the model.
model.compile(optimizer='SGD', 
              loss='categorical_crossentropy',
              metrics=['accuracy'])
# Training the model.
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE, epochs=EPOCHS,
          verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
# Evaluating the model.
test_loss, test_acc = model.evaluate(X_test, Y_test)
print('Test accuracy:', test_acc)


60000 train samples
10000 test samples
Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_layer (Dense)          (None, 128)               100480    
_________________________________________________________________
dense_layer_2 (Dense)        (None, 128)               16512     
_________________________________________________________________
dense_layer_3 (Dense)        (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________
Epoch 1/50
375/375 [==============================] - 1s 4ms/step - loss: 1.4948 - accuracy: 0.6375 - val_loss: 0.7662 - val_accuracy: 0.8347
Epoch 2/50
375/375 [==============================] - 1s 4ms/step - loss: 0.6053 - accuracy: 0.8485 - val_loss: 0.4613 - val_accuracy: 0.8809
Epoch 3/50
375/375 [==============================] - 1

After reduce the number of iterations from 200 to 50 and this time we attain, 97.19% on train data, 96.53% on validation data and  96.51% accuracy on test data.

#Further improving the simple net in TensorFlow with Dropout
Now our baseline is 90.81% on the training set, 91.40% on validation, and 91.18% on test. now we will randomly drop some few values. the idea behind this improvement is that random dropout forces the network to learn redundant patterns that are useful for better generalization

In [ ]:
import tensorflow as tf
import numpy as np
from tensorflow import keras
# Network and training.
EPOCHS = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10   # number of outputs = number of digits
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION
DROPOUT = 0.3
# Loading MNIST dataset.
# Labels have one-hot representation.
mnist = keras.datasets.mnist
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()
# X_train is 60000 rows of 28x28 values; we reshape it to 60000 x 784.
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# Normalize inputs within [0, 1].
X_train, X_test = X_train / 255.0, X_test / 255.0
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# One-hot representations for labels.
Y_train = tf.keras.utils.to_categorical(Y_train, NB_CLASSES)
Y_test = tf.keras.utils.to_categorical(Y_test, NB_CLASSES)
# Building the model.
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(N_HIDDEN,
          input_shape=(RESHAPED,),
          name='dense_layer', activation='relu'))
model.add(keras.layers.Dropout(DROPOUT))
model.add(keras.layers.Dense(N_HIDDEN,
          name='dense_layer_2', activation='relu'))
model.add(keras.layers.Dropout(DROPOUT))
model.add(keras.layers.Dense(NB_CLASSES,
          name='dense_layer_3', activation='softmax'))
# Summary of the model.
model.summary()
# Compiling the model.
model.compile(optimizer='SGD', 
              loss='categorical_crossentropy',
              metrics=['accuracy'])
# Training the model.
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE, epochs=EPOCHS,
          verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
# Evaluating the model.
test_loss, test_acc = model.evaluate(X_test, Y_test)
print('Test accuracy:', test_acc)


60000 train samples
10000 test samples
Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_layer (Dense)          (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_layer_2 (Dense)        (None, 128)               16512     
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_layer_3 (Dense)        (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________
Epoch 1/200
375/375 [==============================] - 2s 4ms/step - loss: 

After running for 200 epochs we attained 97.74% on training data, 97.6% on validation and 97.75% on test that.

![alt text](https://static.packt-cdn.com/products/9781838823412/graphics/Images/B14761_01_25.png)
![alt text](https://static.packt-cdn.com/products/9781838823412/graphics/Images/B14761_01_26.png)

Now i decided to introduce some regularizations and our model performance improve slightly.

In [16]:
import tensorflow as tf
import numpy as np
from tensorflow import keras
from keras.regularizers import l2
# Network and training.
EPOCHS = 200
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10   # number of outputs = number of digits
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION
DROPOUT = 0.3
# Loading MNIST dataset.
# Labels have one-hot representation.
mnist = keras.datasets.mnist
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()
# X_train is 60000 rows of 28x28 values; we reshape it to 60000 x 784.
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# Normalize inputs within [0, 1].
X_train, X_test = X_train / 255.0, X_test / 255.0
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# One-hot representations for labels.
Y_train = tf.keras.utils.to_categorical(Y_train, NB_CLASSES)
Y_test = tf.keras.utils.to_categorical(Y_test, NB_CLASSES)
# Building the model.
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(N_HIDDEN,
          input_shape=(RESHAPED,),
          name='dense_layer', activation='relu'))
model.add(keras.layers.Dropout(DROPOUT))
model.add(keras.layers.Dense(N_HIDDEN,
          name='dense_layer_2', activation='relu'))
model.add(keras.layers.Dropout(DROPOUT))
model.add(keras.layers.Dense(64, input_dim=64,activity_regularizer=l2(0.01)))
model.add(keras.layers.Dense(NB_CLASSES,
          name='dense_layer_3', activation='softmax'))
# Summary of the model.
model.summary()
# Compiling the model.
model.compile(optimizer='SGD', 
              loss='categorical_crossentropy',
              metrics=['accuracy'])
# Training the model.
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE, epochs=EPOCHS,
          verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
# Evaluating the model.
test_loss, test_acc = model.evaluate(X_test, Y_test)
print('Test accuracy:', test_acc)


60000 train samples
10000 test samples
Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_layer (Dense)          (None, 128)               100480    
_________________________________________________________________
dropout_6 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_layer_2 (Dense)        (None, 128)               16512     
_________________________________________________________________
dropout_7 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense (Dense)                (None, 64)                8256      
_________________________________________________________________
dense_layer_3 (Dense)        (None, 10)                650       
Total params: 125,898
Trainable params: 125,898
Non-trainable params: 0
_________